mypy/semanal.py

```diff
@@ -407,6 +406,10 @@ def _visit_func_def(self, defn: FuncDef) -> None:
                             add_symbol = False
                     if add_symbol:
                         self.type.names[defn.name()] = SymbolTableNode(MDEF, defn)
+                if defn.type is not None and defn.name() in ('__init__', '__init_subclass__'):
```

Comment:
Are you sure this works with overloads and decorated methods? Please add tests to verify that.

Also perhaps it should only work inside a class definition?

Reply:
I think one needs similar check at least in `visit_overloaded_func_def`.

Reply:
@gvanrossum It works only inside a class definition.
@gvanrossum  @ilevkivskyi You are right, I will find a way to extend behavior.